numba/np/numpy_support.py

```diff
@@ -424,6 +425,17 @@ def make_specific(outputs, unit):
                     new_outputs.append(out)
             return new_outputs
 
+        def make_datetime_specific(outputs, dt_unit, td_unit):
+            new_outputs = []
+            for out in outputs:
+                if isinstance(out, types.NPDatetime) and out.unit == "":
+                    unit = npdatetime_helpers.combine_datetime_timedelta_units(
+                        dt_unit, td_unit)
```

Comment:
What if unit ends up as `None` as there is no combined type? The result is the `types.NPDatetime()` ctor fails with a key error. Example:
```python
from numba import njit
import numpy as np

@njit
def foo(a, b):
    return a + b

x = np.array([np.datetime64("2011-01-01"),], dtype="datetime64[ns]")
# this is a UFuncTypeError in NumPy without setting a lax casting rule
print(foo.py_func(x, np.timedelta64(1000,'M')))
```

Reply:
Thank you for adding this test. It seems clear to me based on the Numpy output that I should be throwing an exception in this case. I will update the code.

Reply:
I could use a little bit of feedback on how to best throw an exception in this case. I believe that in this situation we should cease compilation since this is a typing error that won't be resolved. However this is a bit less conservative and perhaps raising a Numba extended Error is safer.

However, the Numba extended error won't allow me to maintain the existing error message and I think the no supported implementation is very unclear.

However, the TypeError doesn't allow me to display the loc (and I also don't have enough information to determine it at this stage).

Do you have any preference or advice for the best way to handle this? I don't find either of these particularly satisfying, and I would prefer to be able to show both the code location and my error message.

Reply:
@njriasan I don't think either option is ideal for the reasons noted. Given the current restrictions, I'd probably go with the `TypeError` as you've implemented, at least the reason is clear even if the location is not available. I think it's possible to wire the location through the stack from `typeinfer.py` (it's in the constraint object), but it's quite involved and likely invasive.

Reply:
Okay thank you. I think for this PR I'll opt with the `TypeError` handling.